pkg/controllers/user/alert/manager/prometheus_operator.go

```diff
@@ -85,13 +86,34 @@ func (c *PromOperatorCRDManager) SyncPrometheusRule(promRule *monitoringv1.Prome
 		return fmt.Errorf("get namespace %s for prometheus rule %s:%s failed, %v", ns.Name, promRule.Namespace, promRule.Name, err)
 	}
 
+	//sort before compare
+	var sortedGroups []monitoringv1.RuleGroup
+	for _, v := range promRule.Spec.Groups {
+		sortedRules := v.Rules
+		sort.Slice(sortedRules, func(i, j int) bool {
+			return sortedRules[i].Labels["rule_id"] < sortedRules[j].Labels["rule_id"]
+		})
+		group := v.DeepCopy()
+		group.Rules = sortedRules
+		sortedGroups = append(sortedGroups, *group)
+	}
+
+	sort.Slice(sortedGroups, func(i, j int) bool {
+		return sortedGroups[i].Name < sortedGroups[j].Name
+	})
+	promRule.Spec.Groups = sortedGroups
+
 	old, err := c.prometheusRules.GetNamespaced(promRule.Namespace, promRule.Name, metav1.GetOptions{})
 	if err != nil {
 		if !apierrors.IsNotFound(err) {
 			return fmt.Errorf("get prometheus rule %s:%s failed, %v", promRule.Namespace, promRule.Name, err)
 		}
-		if old, err = c.prometheusRules.Create(promRule); err != nil && !apierrors.IsAlreadyExists(err) {
```

Comment:
I don't think you need to change the logic here. You should `return nil` after create success.